rust/windows-client/src-tauri/src/client/gui.rs

```diff
@@ -377,117 +407,142 @@ impl Controller {
             token.clone(),
             device_id,
             CallbackHandler {
-                ctlr_tx,
                 logger: Some(logger),
+                notify_controller,
+                resources,
             },
         )?)
     }
 }
 
+// TODO: After PR #2960 lands, move some of this into `impl Controller`
 async fn run_controller(
     app: tauri::AppHandle,
     mut rx: mpsc::Receiver<ControllerRequest>,
     logging_handles: client::logging::Handles,
     advanced_settings: AdvancedSettings,
+    notify_controller: Arc<Notify>,
 ) -> Result<()> {
-    let mut controller = Controller::new(app.clone(), logging_handles, advanced_settings)
-        .await
-        .context("couldn't create Controller")?;
+    let mut controller = Controller::new(
+        app.clone(),
+        logging_handles,
+        advanced_settings,
+        notify_controller,
+    )
+    .await
+    .context("couldn't create Controller")?;
 
     tracing::debug!("GUI controller main loop start");
 
-    while let Some(req) = rx.recv().await {
-        match req {
-            Req::ExportLogs(file_path) => settings::export_logs_to(file_path).await?,
-            Req::GetAdvancedSettings(tx) => {
-                tx.send(controller.advanced_settings.clone()).ok();
-            }
-            Req::SchemeRequest(url) => {
-                if let Ok(auth) = parse_auth_callback(&url) {
-                    tracing::debug!("setting new token");
-                    let entry = keyring_entry()?;
-                    entry.set_password(auth.token.expose_secret())?;
-                    controller.connlib_session = Some(Controller::start_session(
-                        &controller.advanced_settings,
-                        controller.ctlr_tx.clone(),
-                        controller.device_id.clone(),
-                        &auth.token,
-                        controller.logging_handles.logger.clone(),
-                    )?);
-                    controller.session = Some(Session {
-                        actor_name: auth.actor_name,
-                        token: auth.token,
-                    });
-                } else {
-                    tracing::warn!("couldn't handle scheme request");
-                }
-            }
-            Req::SignIn => {
-                // TODO: Put the platform and local server callback in here
-                tauri::api::shell::open(
-                    &app.shell_scope(),
-                    &controller.advanced_settings.auth_base_url,
-                    None,
-                )?;
-            }
-            Req::SignOut => {
-                keyring_entry()?.delete_password()?;
-                if let Some(mut session) = controller.connlib_session.take() {
-                    // TODO: Needs testing
-                    session.disconnect(None);
+    loop {
+        controller.notify_controller.notified().await;
+        tracing::debug!("controller woke up");
+
+        match rx.try_recv() {
```

Comment:
this is what I mean, it seems like you're re-implementing the mpsc channel `recv().await` I don't think there's any benefit.